pike/test/copychunk.py

```diff
@@ -428,3 +453,165 @@ def test_neg_dst_is_a_dir(self):
+                # for different file handler in same session
+                self.assertNotEqual(resume_key_1_1, resume_key_2_1)
+                # for different file hander and different session
+                self.assertNotIn(resume_key_1_1, rk_list)
+                self.assertNotIn(resume_key_2_1, rk_list)
+            except Exception as e:
+                raise AssertionError("resume key check fail", e)
+            else:
+                rk_list += [resume_key_1_1, resume_key_2_1]
+
+    def test_multiple_resume_key(self):
+        """
+        test multiple sessions requesting resume_key for same
+        file, also test within one session, multiple requests
+        for same file
+        """
+        self.generic_multiple_resume_key_test_case(sess_num=3)
+
+    def generic_negative_resume_key_test_case(self, resume_key, exp_error=None):
+        """
+        negative test for invalid resume key handling case
+        """
+        src_filename = "src_copy_chunk_rk.txt"
+        dst_filename = "dst_copy_chunk_rk.txt"
+
+        chunks = [(0, 0, len(SIMPLE_CONTENT))]
+        fh_src, fh_dst = self._open_src_dst(
+            src_filename, dst_filename)
+
+        if exp_error is None:
+            exp_error = pike.ntstatus.STATUS_SUCCESS
+        try:
+            with self.assert_error(exp_error):
+                result = self.chan.copychunk(
+                    fh_src, fh_dst, chunks, resume_key)
+        finally:
+            self.chan.close(fh_src)
+            self.chan.close(fh_dst)
+
+    def test_other_resume_key(self):
+        """
+        negative test for multiple sessions acquiring resume key
+        for same file one session uses other session's resume key
+        to perform the server side copy
+        """
+        self._prepare_source_file()
+        chan_2, tree_2 = self.tree_connect()
+        src_filename = "src_copy_chunk_rk.txt"
+        fh_src_2 = chan_2.create(tree_2,
+                                 src_filename,
+                                 access=access_rwd,
+                                 share=share_all).result()
+        self.other_chan_list.append(
+            {"channel": chan_2, "tree": tree_2, "filehandles": [fh_src_2]})
+        other_key = chan_2.resume_key(fh_src_2)[0][0].resume_key
+
+        self.generic_negative_resume_key_test_case(
+            other_key, exp_error=pike.ntstatus.STATUS_OBJECT_NAME_NOT_FOUND)
+
+    def test_bogus_resume_key(self):
+        """
+        negative test for single session does not acquire resume_key
+        but generate one bogus resume_key to perform the server side copy
+        """
+        self._prepare_source_file()
+        import array
+        bogus_key = array.array('B', [255] * 24)
+        self.generic_negative_resume_key_test_case(
+            bogus_key, exp_error=pike.ntstatus.STATUS_OBJECT_NAME_NOT_FOUND)
+
+    def generic_ssc_write_test_case(self, block, number_of_chunks, total_offset=0):
```

Comment:
the purpose of this function is identical to generic_ssc_test_case but using different dst permissions and different copy chunk command. consider parameterizing the original `generic_ssc_test_case` to include these variations, or consider abstracting some of the duplicated logic into helper functions that both generic test cases can use

Reply:
I try to merge them into one common generic test case, please check it is acceptable.

Reply:
when i'm running it now, i hit a different problem

```
======================================================================
ERROR: test_copy_write_small_file (pike.test.copychunk.TestServerSideCopy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pike/test/copychunk.py", line 241, in test_copy_write_small_file
    self.generic_ssc_test_case(block, num_of_chunks, write_flag=True)
  File "pike/test/copychunk.py", line 183, in generic_ssc_test_case
    result = self.chan.copychunk(fh_src, fh_dst, chunks, write_flag)
  File "pike/model.py", line 1777, in copychunk
    write_flag).parent.parent)[0]
  File "pike/model.py", line 860, in transceive
    return map(Future.result, self.submit(req))
  File "pike/model.py", line 733, in _prepare_outgoing
    result = req.parent.serialize()
  File "pike/core.py", line 473, in serialize
    self.encode(cursor)
  File "pike/core.py", line 462, in encode
    self._encode(cur)
  File "pike/netbios.py", line 61, in _encode
    child.encode(cur)
  File "pike/core.py", line 462, in encode
    self._encode(cur)
  File "pike/smb2.py", line 196, in _encode
    self._command.encode(cur)
  File "pike/core.py", line 462, in encode
    self._encode(cur)
  File "pike/smb2.py", line 3167, in _encode
    self.ioctl_input.encode(cur)
  File "pike/core.py", line 462, in encode
    self._encode(cur)
  File "pike/smb2.py", line 3260, in _encode
    cur.encode_bytes(self.source_key)
  File "pike/core.py", line 203, in encode_bytes
    size = len(val)
TypeError: object of type 'bool' has no len()
```

this is against windows server 2016